### DISCUSSION TITLE
fasta to tab to ... ?

### DISCUSSION THREAD

**Discussion by skreling** [sr320/course-fish546-2021#75] [Q&A] [answered]
So I successfully converted the fasta into a .tab file, but there were secondary delimiters in there that I wanted in separate columns, so I removed the tabs and replaced all tabs with ";" so that everything was semicolon delimited. However, in this process somewhere along the line there was a mismatch in column names, so that everything had values in not the correct columns, but not in a patterned way that I could just shuffle things around via code in R. Is there a way to pull the parts of the header from fasta out in a less crude way than I've been attempting to? 

Link to original tab file: https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Results/Wolf/wolf_results_table.tab
Link to next attempt in removing tabs: https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Results/Wolf/wolf_results_table2.tab
Link to attempt to manually turn it to comma delimited (also converted this to a csv after and R still was not a fan): https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Results/Wolf/wolf_results_table3.tab

**Comment by sr320** [comment]
This has to do with fact that there are not equal columns if you separate by `;` found in "original" 

See https://d.pr/i/SMF6Sv

Why are doing this? Is there another way you can describe your end goal?

**Reply by skreling** [reply]
Yes I realized this, but I was wondering if there was a way to call out the "TAXID=" or something to that regard for each entry so that you could form a table with the information provided. I mean I can manually turn this into a table because there's not that many taxa represented in the diet sample, but that's not ideal. I really just want number of reads per taxa which are just two columns (if I could get them to format correctly) in the data

**Reply by kubu4** [reply]
Here's the quick way to pull out (using the shell) just rows with "TAXID":

`grep "TAXID" wolf_results_table3.tab`

**Reply by kubu4** [reply]
> I really just want number of reads per taxa

Could you load your `wolf_results_table3.tab` and tell us which columns this info is in? It's not obvious to me which column(s) contain this info.

**Reply by skreling** [reply]
So it's not in a specific column rn that's kind of the issue but you'll see the 'COUNT=' and the 'Scientific_Name=' headers strewn in there and that's essentially what I want to extract

**Comment by kubu4** [comment]
> R still was not a fan)

Could you elaborate on this a bit? Did you get an error message? If yes, could you please share what message you received?

I was able to read your final file in without issue in R:

```R
df = read.csv(file = "Downloads/class_546/Sam-Metabarcoding/Results/Wolf/wolf_results_table3.tab", header = FALSE)
```

![image](https://user-images.githubusercontent.com/4514104/107772607-4b3ae080-6cf1-11eb-8899-06343dc19a0e.png)


Along these lines, if you're able, it'd be helpful if you could tell us how you generated (or where you got) the FastA file. I'm going under the assumption that you used OBITools to generate the FastA? If yes, I see that the example tutorial ends with a FastA conversion to a tab-delimited file using `obitab`. Any reason you didn't (or weren't able to?) use `obitab`?

Finally, regarding Steven's observation that three are a different number of columns in some of the rows, you could subset your data to only include rows which have only the info you're truly interested in looking at. For example, you could subset your data to only include rows that have a `TAXID` value

**Reply by skreling** [reply]
No R error! And yes you can read the third one in R fine, but if you look at the columns, there were uneven numbers of ";" to delimit different columns, so the columns in that file are essentially all mixed up. 

I did use the OBITools to generate the fasta and I did get the resulting obitab file, but that only gives you the "HELIUM_000" in a single column and nothing else which seems a bit useless. But I'll try again and see what happens, maybe I just messed up

**Comment by kubu4** [comment, marked as answer]
If you could share your workflow/commands and resulting files that you used, that would allow us to help troubleshoot a bit better.

Side note - do you have any tips on getting OBITools installed? I tried during class last week and was unsuccessful.

**Reply by skreling** [reply]
Problem is I couldn't get the virtual environment to work in Jupyter so it was all command line and thus I do not have the exact workflow I used, but it was very similar to : https://git.metabarcoding.org/obitools/obitools3/wikis/Wolf-tutorial-with-the-OBITools3

Installation: Took me a while to figure it out, but the original obi tools relied on specifically python 2.7 which is no longer supported, so they upgraded Obitools, but did not make that easy to find, so if you were trying the older package that might have been it. Heres a link to the new version. They have an installation page on their wiki that worked for me: https://git.metabarcoding.org/obitools/obitools3/wikis/Installing-the-OBITools3


It's honestly fine, though I'll just figure this out!

**Reply by kubu4** [reply]
Dang! This software is _slick_! It even has a history command `obi history` that will list all of the commands run! Also, it produces a visualization of the entire pipeline that was run?!?! (`dot -Tx11 wolf.dot`) Awesome!

Now that you helped me get it installed, I should be able to help troubleshoot a bit more accurately.

**Reply by kubu4** [reply]
> I couldn't get the virtual environment to work in Jupyter

Here's how I got it to work in Jupyter, if you're curious:

1. Create a new conda environment for obitools3

`conda create --name obitoolsenv`

2. Activate that environment:

`conda activate obitoolsenv`

3. Install `pip`:

`conda install pip`

4. Install obitools3 dependencies:

`pip install --upgrade pip setuptools wheel Cython`

5. Install obitools3:

`pip install --pre --upgrade OBITools3`

6. Install Jupyter Lab:

`conda install -c conda-forge jupyterlab`

7. Start Jupyter Lab:

`jupyter-lab`

The, use your Jupyter Notebook! Remember, commands will need to be prefaced with an `!` (or, use `%%bash` on the first line of each cell) in order to use programs that are accessed via the command line.

**Reply by kubu4** [reply]
![image](https://user-images.githubusercontent.com/4514104/107811718-763e2800-6d23-11eb-9e10-db3a632e870b.png)

**Reply by skreling** [reply]
This was really helpful thank you! Yes obi tools does seem really powerful and useful for metabarcoding! Seems to be pretty standard across all types of metabarcoding as well!

**Comment by kubu4** [comment]
@skreling Did you post a problem earlier today, regarding accidentally deleting your obitools conda environment? I have it in my email, but when I follow the link to this discussion, it doesn't seem to exist.

**Reply by skreling** [reply]
Yes! Sorry I figured out a work around almost immediately after posting it so I just deleted it.

**Reply by kubu4** [reply]
Ah! Gotcha.

In the future, if you could leave posts up, that'd be great. It helps serve as a resource for other people who might encounter the same/similar problems.

Most importantly, though, glad you got it resolved! Nice work!
